### PR TITLE
Arde/fsdp activation checkpointing

### PR DESCRIPTION
**Pull request by arde171**
# What does this PR do?
Currently, HF Trainer didn't support FSDP activation checkpointing. This PR provides support to FSDP activation checkpointing.
 Please see the details about the FSDP activation checkpointing [here](https://pytorch.org/blog/scaling-multimodal-foundation-models-in-torchmultimodal-with-pytorch-distributed/#activation-checkpointing).
I saw the improvement in training performance for the large LLM models (e.g., LLAMA 70B) with FSDP activation checkpointing as compared to the existing gradient_checkpointing option. It's easy to enable FSDP activation_checkpointing. 

we just need to add `"activation_checkpointing": "True"` to enable the FSDP `activation_checkpointing` as shown in below example `fsdp_config.json` file.

fsdp_config.json
```
{
  "transformer_layer_cls_to_wrap": ["LlamaDecoderLayer"],
  ...
  "activation_checkpointing": "True"
}
``` 
Please see the below PR for more details about FSDP activation checkpointing in accelerate repo:
PR: https://github.com/huggingface/accelerate/pull/1891

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -471,6 +471,10 @@ def __init__(
             if self.args.fsdp_config.get("limit_all_gathers", False):
                 self.limit_all_gathers = True
 
+            self.activation_checkpointing = False
+            if self.args.fsdp_config.get("activation_checkpointing", False):
+                self.activation_checkpointing = True
+
```
Comment (pacman100) [inline src/transformers/trainer.py:477]: not required, pleasec remove these lines

**Comment by ArthurZucker** [comment]
Cc @pacman100 if you think this is relevant 🤗

**Review by pacman100 — COMMENTED** [review]
Thank you @arde171 for adding this. Please raise an error if both `activation_checkpointing` in FSDP config and training arg `gradient_checkpointing` are set to `True`. The error should mention that both can't be set to True and to use FSDP's checkpointing logic when using FSDP.

File: src/transformers/trainer.py (hunk 1 above)
Comment (arde171) [reply src/transformers/trainer.py]: Done.

**Review by pacman100 — APPROVED** [review]
Thank you @arde171!

**Pull request merged**